carbon-design-system/carbon
# fix(tooltip): ensure esc closes tooltip, avoid bubbling beyond the body

Closes #10204 

#### Changelog

**New**

- test(tooltip): add test to cover "<kbd>esc</kbd> to close" functionality

**Changed**

- fix(tooltip): remove global/document <kbd>esc</kbd> keyDown handler, replace it with scoped onKeyDown handler on tooltip body
- docs(tooltip): remove interactive elements from the tooltip body in the default story
- docs(modal): add a tooltip to the WithStateManager story to more easily test tooltips inside an element with it's own "<kbd>esc</kbd> to close" handler.

#### Testing / Reviewing

I'm sorry this has quite a few steps to fully test the changes here:

- In the `carbon-components-react` storybook deploy preview:
  - Ensure tooltip without interactive body contents can be closed via <kbd>esc</kbd>
  - Ensure tooltip with interactive body contents can be close via <kbd>esc</kbd>
  - Ensure Modal WithStateManager tooltip can be closed via <kbd>esc</kbd> and _does not close the modal_. A _second_ press of <kbd>esc</kbd> should close the modal.
- Pull this down to test locally - we need to ensure the same steps from above work w/ React v16
  - Modify the react deps in the root `package.json` to be 16.8.6 ([here is a reference](https://github.com/carbon-design-system/carbon/pull/10029/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519) from the PR updating it to v17)
  - Modify the react deps in the `packages/react/package.json` to be 16.8.6
  - `yarn install` at the root
  - `yarn start` the `packages/react` storybook
  - You can validate the storybook is using React v16 by adding a `console.log(React.version);` (anywhere, in a component, etc) - it should output to the console `v16.8.6`
  - Redo the tests that you did on the deploy preview:
    - Ensure tooltip without interactive body contents can be closed via <kbd>esc</kbd>
    - Ensure tooltip with interactive body contents can be close via <kbd>esc</kbd>
    - Ensure Modal WithStateManager tooltip can be closed via <kbd>esc</kbd> and _does not close the modal_. A _second_ press of <kbd>esc</kbd> should close the modal.
